Implement lock destruction feature + cleanup in locking-related code

hlef opened the pull request:
This PR contains three commits:

- First one fixes some stale comments which I found somewhat confusing
- Second one implements lock destruction (see issue #168)
- Last one does some cleanup in the locking tests for coherence and consistency

It would be nice to implement `test_destruct_flag_lock_acquire` with templating, unfortunately we are hitting a compiler bug which prevents us from doing that. Let's add templating to that function once the compiler bug is reported and fixed.

davidchisnall commented on sdk/lib/locks/locks.cc:
```diff
@@ -136,14 +132,33 @@ namespace
+
+		/**
+		 * Set the destruction bit in the flag lock word and wake
+		 * waiters. Assumes that the lock is hold by the caller.
```
```suggestion
		 * waiters. Assumes that the lock is held by the caller.```

hlef commented:
@microsoft-github-policy-service agree company="SCI Semiconductor Ltd."

hlef commented:
Thanks for the review @davidchisnall !

I have updated the PR with fixes to your comments. Two things to note:
- I did not do exactly what you suggested with `TEST(lock.lock())` in the async, because the `lock.lock()` API does not return a bool. To reach a compromise, I added an assert on `try_lock()` before `lock.lock()`. Not perfect, but better than nothing?
- I did not add an assert for the API misuse `lock.unlock()`. Would you like me to add the assert? If so, what would be the preferred approach to reset the lock?

hlef commented:
For the record, the compiler bug which prevents us from using templating in the second test is this one: https://github.com/CHERIoT-Platform/llvm-project/issues/16

davidchisnall commented on sdk/lib/locks/locks.cc:
```diff
@@ -63,6 +52,17 @@ namespace
+
+				// Do not attempt to lock if the destruct mode
+				// flag is set. Note that the lock is in
+				// destruction mode iff it is already locked,
```
I don't think that's the reason: we are not setting the locked-for-destruction bit in our expected value, so the CAS will always fail if the lock-for-destruction bit is set.  Being locked for destruction is unusual (it should happen exactly once for any lock) and so we don't bother to check for it unless we've failed to acquire the lock.

davidchisnall commented on sdk/lib/locks/locks.cc:
```diff
@@ -136,14 +136,33 @@ namespace
+		 * Set the destruction bit in the flag lock word and wake
+		 * waiters. Assumes that the lock is held by the caller.
+		 *
+		 * Note: This does not check that the lock is owned by the
```
It might be nice to have an assert that does check.  There's a `Debug::Assert` variant that takes a lambda as the argument and so will not compile the check unless we've turned on lock debugging.

hlef replied:
I agree, these would be nice to have!

I suggest to add them in a separate commit. Currently, most methods in this file make this assumption without any assert. The commit which adds the assert could do so on all methods at once.

davidchisnall commented on tests/locks-test.cc:
```diff
@@ -65,6 +65,81 @@ namespace
+
+		// Create an async that tries to grab the lock
+		async([&]() {
+			lock.lock();
```
The lock with no timeout will block this thread forever if the wake fails.  Subsequent tests will then deadlock.

hlef commented:
@davidchisnall

> The lock with no timeout will block this thread forever if the wake fails. Subsequent tests will then deadlock.

I don't think that this is still a problem in the current version. If the wake fails, we are guaranteed to fail `TEST(modified == true, ...)`, which will prevent other tests from being executed anyways.

hlef commented:
@davidchisnall I updated the PR to address your comment on "Do not attempt to lock if the destruct mode...".

Regarding the assert, I suggest to do that as part of another PR since most of the code in `locks.cc` would benefit from this.

davidchisnall commented:
> @davidchisnall
> 
> > The lock with no timeout will block this thread forever if the wake fails. Subsequent tests will then deadlock.
> 
> I don't think that this is still a problem in the current version. If the wake fails, we are guaranteed to fail `TEST(modified == true, ...)`, which will prevent other tests from being executed anyways.

If any of the asserts fire on the main thread, they will force unwind out and the unlock will not be reached. The next test will then run. Try it and see.

hlef commented:
> > @davidchisnall
> > > The lock with no timeout will block this thread forever if the wake fails. Subsequent tests will then deadlock.
> > 
> > 
> > I don't think that this is still a problem in the current version. If the wake fails, we are guaranteed to fail `TEST(modified == true, ...)`, which will prevent other tests from being executed anyways.
> 
> If any of the asserts fire on the main thread, they will force unwind out and the unlock will not be reached. The next test will then run. Try it and see.

Alright, I have reworked the code a bit, now it should be good?

hlef commented:
Thank you for the comments, pushed a new version to address them.

davidchisnall reviewed (approved):
Looks great, thank you!
